Allow tooling authors to provide custom postcss factory through options

phated opened the pull request:
<!--
  HOLY CRAP a Pull Request. We ❤️ those!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Please place an x (no spaces!) in all [ ] that apply
-->

This PR contains a:

- [ ] **bugfix**
- [x] new **feature**
- [ ] **code refactor**
- [x] **test update** <!-- if bug or feature is checked, this should be too -->
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

Ref #510 

I know @alexander-akait said it is out-of-scope, but I needed to write the code anyway. So I figured I'd make the code suitable to upstream and submit it anyway—under the assumption this will be closed.

I'll be publishing my fork as `@phated/postcss-loader` for our use in Storybook until we can make the 7.0 breaking release—at which time, we'll be able to require users to rely on peerDeps and switch back to this upstream project.

Cheers! 🚀 

<!--
  Please explain the motivation or use-case for your change.
  What existing problem does the PR solve?
  If this PR addresses an issue, please link to the issue.
-->

### Breaking Changes

<!--
  If this PR introduces a breaking change, please describe the impact and a
  migration path for existing applications.
-->

### Additional Info

alexander-akait reviewed (commented):
Can you provide example of usage? Do you want to use `postcss@7` or `postcss@8`? Why not validate it on boilerplate side? In theory I can add this as the migration options (and when we drop postcss@7 support we remove it)

phated commented:
Currently, the way a user consumes Storybook looks (more-or-less) like:

```
/users-project-root
├─┬ node_modules
│ └─┬ storybook
│   ├── postcss-loader
│   └── postcss@^7
└── storybook-config-file
```

The internally-included `postcss@^7` dependency cannot be removed before the 7.0 released, and users are using it implicitly.

However, a new option can be exposed that allows users to specify a custom postcss factory that has priority over the implicit dependency using this feature. So they would do:

```
/users-project-root
├─┬ node_modules
│ ├─┬ storybook
│ │ ├── postcss-loader
│ │ └── postcss@^7
│ └── postcss@^8 <-- add postcss 8 as top-level dep
└── storybook-config-file <-- add require("postcss")
```

alexander-akait reviewed (commented):
Okay, let's rename `customPostcss` to `implementation`, we have the same option in `sass-loader` (https://github.com/webpack-contrib/sass-loader/blob/master/src/options.json#L4)

phated commented:
@alexander-akait great thought! I've updated to be more consistent with sass-loader. Do you have any recommendations for getting the snapshots to succeed consistently across node versions?

alexander-akait commented on test/cjs.test.js:
```diff
@@ -1,3 +1,4 @@
+// eslint-disable-next-line import/no-named-as-default, import/no-named-as-default-member
```
I think it is global eslint problem, we don't have this rules

alexander-akait commented on src/index.js:
```diff
@@ -1,7 +1,7 @@
 import { validate } from "schema-utils";
 
-import postcss from "postcss";
+import ourPostcss from "postcss";
```
Let's return to `postcss`

phated replied:
The problem I ran into was naming the local variable `postcss` too. I can change that and put this back to `postcss`

alexander-akait commented on README.md:
```diff
@@ -593,6 +593,38 @@ module.exports = {
+### `implementation`
+
+Type: `Function`
+Default: `undefined`
```
`require('postcss')`

alexander-akait commented:
Also can you accept CLA?

phated commented:
I think the CLA app is bugged

<img width="870" alt="Screen Shot 2021-01-20 at 10 35 00 AM" src="https://user-images.githubusercontent.com/992373/105212903-45474a80-5b0b-11eb-8ce1-e7963f4a540f.png">

phated commented:
@alexander-akait I addressed your comments and squashed the commits. I removed my package-lock updates and other things that made the tests pass locally for me, I think those should be fixed outside this feature.

alexander-akait commented:
Thanks
